management/operations.go

```diff
@@ -2,67 +2,69 @@ package management
+type Operation struct {
+	XMLName        xml.Name `xml:"http://schemas.microsoft.com/windowsazure Operation"`
 	ID             string
-	Status         string
+	Status         OperationStatus
 	HttpStatusCode string
 	Error          AzureError
 }
 
-//getOperationStatus gets an operation given the operation ID.
-func (client *Client) getOperationStatus(operationId string) (*operation, error) {
+type OperationStatus string
+
+const (
+	OperationStatusInProgress OperationStatus = "InProgress"
+	OperationStatusSucceeded  OperationStatus = "Succeeded"
+	OperationStatusFailed     OperationStatus = "Failed"
+)
+
+// OperationId is assigned by Azure API and can be used to look up the status of an operation
+type OperationId string
+
+// getOperationStatus gets an operation given the operation ID.
+func (client *Client) GetOperationStatus(operationId OperationId) (Operation, error) {
+	operation := Operation{}
 	if operationId == "" {
-		return nil, fmt.Errorf(errParamNotSpecified, "operationId")
+		return operation, fmt.Errorf(errParamNotSpecified, "operationId")
 	}
 
-	operation := new(operation)
-	url := "operations/" + operationId
+	url := fmt.Sprintf("operations/%s", operationId)
 	response, azureErr := client.SendAzureGetRequest(url)
 	if azureErr != nil {
-		return nil, azureErr
+		return operation, azureErr
 	}
 
-	err := xml.Unmarshal(response, operation)
+	err := xml.Unmarshal(response, &operation)
 	if err != nil {
-		return nil, err
+		return operation, err
 	}
 
 	return operation, nil
 }
 
-//waitAsyncOperation blocks until the operation with the given operationId is
-//no longer in the InProgress state. If the operation was successful, nothing is
-//returned, otherwise an error is returned.
-func (client *Client) WaitAsyncOperation(operationId string) error {
-	if operationId == "" {
-		return fmt.Errorf(errParamNotSpecified, "operationId")
-	}
-
-	status := "InProgress"
-	operation := new(operation)
-	err := errors.New("")
-	for status == "InProgress" {
-		time.Sleep(2000 * time.Millisecond)
-		operation, err = client.getOperationStatus(operationId)
+// WaitAsyncOperation blocks until the operation with the given operationId is
+// no longer in the InProgress state. If the operation was successful, nothing is
+// returned, otherwise an error is returned.
+func (client *Client) WaitAsyncOperation(operationId OperationId) error {
+	for {
+		operation, err := client.GetOperationStatus(operationId)
 		if err != nil {
 			return err
 		}
 
-		status = operation.Status
-	}
-
-	if status == "Failed" {
-		return fmt.Errorf("Azure operation %s failed. Code: %s, Message: %s", operationId, operation.Error.Code, operation.Error.Message)
+		if operation.Status == OperationStatusFailed {
+			return fmt.Errorf("Azure operation %s failed. Code: %s, Message: %s", operationId, operation.Error.Code, operation.Error.Message)
+		} else if operation.Status != OperationStatusInProgress {
+			return nil
+		}
+		time.Sleep(2000 * time.Millisecond)
```

Comment:
you delay every successful response 2 seconds anyway. also this is too high. most SDKs use default 30s.